Add pyadb keygen command to generate keypairs

joeleong opened the pull request:
Should address #95 and #131 

* (I think) I reimplemented the custom public key structure from AOSP.
* Add `pyadb keygen` command to mimic `adb keygen`

Note: I did update the dependencies since the pubkey code uses `pycryptodome` and `six`.
 I was having problems similar to #109, so I think you need `M2Crypto` or `rsa` and `pycryptodome` (2 libraries). There is probably a way to implement the pubkey stuff with the other crypto libraries but I didn't bother. Happy to change things if necessary.

Halastra commented:
Let's help this thing reach master!
1. I'm not experienced in creating proper test cases, but I'm willing to try.
2. I have a minor improvement suggestion for struct usage in decode_pubkey (will post proposed patch soon).
3. I'd also like to try and make this work with the cryptography library (recently supported by this project).

Hopefully I'll be able to do it soon.
Thanks for your great work!

Halastra commented on adb/android_pubkey.py:
```diff
@@ -0,0 +1,157 @@
+  # (3 * sizeof(uint32_t) + 2 * ANDROID_PUBKEY_MODULUS_SIZE)
+
+# Size of the RSA modulus in words.
+ANDROID_PUBKEY_MODULUS_SIZE_WORDS = (ANDROID_PUBKEY_MODULUS_SIZE // 4)
```
Use `size_in()`:
```suggestion
ANDROID_PUBKEY_MODULUS_SIZE_WORDS = size_in(
    ANDROID_PUBKEY_MODULUS_SIZE, # Original size (in bytes)
    struct.calcsize('L') # We want the size in words
)
```

Halastra commented on adb/android_pubkey.py:
```diff
@@ -0,0 +1,157 @@
+import Crypto.Util
+import Crypto.PublicKey.RSA
+
+
```
Use a function for size translation:
```suggestion
def size_in(x, s):
    """
    Get the amount of s-sized blocks required to hold x.
    @param x: original size
    @type x: int
    @param s: desired block size
    @type s: int
    @rtype: int
    """
    return ((x - 1) // s + 1)
```

Halastra commented on adb/android_pubkey.py:
```diff
@@ -0,0 +1,157 @@
+
+
+# Size of an RSA modulus such as an encrypted block or a signature.
+ANDROID_PUBKEY_MODULUS_SIZE = (2048 // 8)
```
```suggestion
ANDROID_PUBKEY_MODULUS_SIZE_BITS = 2048
ANDROID_PUBKEY_MODULUS_SIZE = size_in(
    ANDROID_PUBKEY_MODULUS_SIZE_BITS, #  Original size (in bits)
    8)  # Desired group size (8 bits per byte)
```

Halastra commented on adb/android_pubkey.py:
```diff
@@ -0,0 +1,157 @@
+def _to_bytes(n, length, endianess='big'):
+    """partial python2 compatibility with int.to_bytes
+    https://stackoverflow.com/a/20793663"""
+    if six.PY2:
```
Verify `endianness` argument:
```assert endianess in ('little', 'big')```

Halastra commented on setup.py:
```diff
@@ -62,6 +55,8 @@
     install_requires = [
         'libusb1>=1.0.16',
+        'pycryptodome',
+        'six',
```
Remove unnecessary requirement:
```suggestion
```

Halastra commented:
Hi. I completely agree :) 
Let me explain, and then tell me what you think. Maybe it's redundant after all.

The `size_in` function lets you convert sizes between different units.
In our case, we convert 2048 bits to 256 bytes and then 64 words.

But what if the modulus length is not a multiple of 32 or 8 bits?
e.g. 2047 bits would still require 256 bytes, and 2040 bits would require 255 bytes which are still 64 words.

The `size_in` is a simple formula which handles these cases. 
I don't really know cryptography really well. Perhaps these cases are irrelevant by design.

Also agreed on the tests, probably most important. I'll try to get to those soon 👍

joeleong commented:
I rebased and incorporated most of your suggestions in 0791b9. 
I think I'd prefer to skip the `size_in` suggestion for now, just because they're constants.
I also forgot to include your suggestion about asserting the endianess argument, but I don't think that's a big deal since it defaults to little if gibberish is passed and its a local function anyway.

Let me know what you think!

Halastra commented:
@googlebot I consent

Halastra commented:
Well that took longer than expected.. but here goes.
A few quick notes:
- This test won't actually run on Python 3+ because of #159 (I will address this issue in #157 or a new PR)
- I'm misusing this platform because I don't know how to add commits to this PR. My apologies.
- This test is simple and naive, but should be enough to cover most of your added functionality.
- Let's hope `TemporaryDirectory` is allowed to run on the CI server :0

Add the following code under `test/test_cryptography.py`:
```
import unittest
from adb import sign_cryptography
from adb import android_pubkey
from tempfile import TemporaryDirectory
from os import path

class CryptographyTest(unittest.TestCase):
    def testKeygenAndSigner(self):
        private_key_file_name = 'adbkey'
        with TemporaryDirectory() as temp_dir:
            private_key_file_path = path.join(
                temp_dir, 
                private_key_file_name
            )
            android_pubkey.keygen(private_key_file_path)
            signer = sign_cryptography.CryptographySigner(private_key_file_path)
```

Regarding your previous notes:
- I really like your work!
- I might be tempted to add the `size_in` function in a future PR.
- Same goes for the endianess (+ better default to 'little', to match `bytes.fromhex`)

Please add the above changes. I will review and approve them (promise!).

joeleong commented:
Hey thanks for your suggestions!

I haven't really been keeping up with this project, particularly #167 and all the stuff @JeffLIrion has been contributing, so it is a little unclear to me what the best path forward is. It's very possible his PR's could supersede this one.

I'm happy to incorporate your suggestions to this PR either directly or you can maybe submit a PR to https://github.com/joeleong/python-adb/tree/add-custom-pubkey-encoding and I'll push your commits here.

Halastra commented:
You are right in your observations. His work is indeed an improvement in many areas!
But since there's still no apparent USB support in `adb-shell`, I'll be using this package.

So I guess I'll submit a PR to your repository as you suggested.

JeffLIrion commented:
This pull request contains @joeleong's key generation code as well as tests: https://github.com/google/python-adb/pull/170

I also copied it over to `adb-shell` (and attributed it to @joeleong), along with tests. 

* [`keygen.py`](https://github.com/JeffLIrion/adb_shell/blob/master/adb_shell/auth/keygen.py)
* [`test_adb_device.py`](https://github.com/JeffLIrion/adb_shell/blob/f6bacb8ae2d978b257656995e91080ddf2d6e0c0/tests/test_adb_device.py#L99)
* [`test_keygen.py`](https://github.com/JeffLIrion/adb_shell/blob/master/tests/test_keygen.py)
* [Coverage of `keygen.py`](https://coveralls.io/builds/27227613/source?filename=adb_shell%2Fauth%2Fkeygen.py)
